bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/internal/OpenThermGatewaySocketConnector.java

```diff
@@ -0,0 +1,227 @@
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.openthermgateway.internal;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.PrintWriter;
+import java.net.InetSocketAddress;
+import java.net.Socket;
+import java.util.AbstractMap;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import org.eclipse.smarthome.core.types.State;
+import org.openhab.binding.openthermgateway.handler.TypeConverter;
+
+/**
+ * The {@link OpenThermGatewaySocketConnector} is responsible for handling the socket connection
+ *
+ * @author Arjen Korevaar - Initial contribution
+ * @author Arjan Mels - Improved robustness by re-sending commands, handling all message types (not only Boiler)
+ */
+public class OpenThermGatewaySocketConnector implements OpenThermGatewayConnector {
+    private static final int COMMAND_RESPONSE_TIME = 100;
+    private static final int COMMAND_TIMEOUT = 5000;
+    private OpenThermGatewayCallback callback;
+    private String ipaddress;
+    private int port;
+
+    private Socket socket;
+    private BufferedReader reader;
+    private PrintWriter writer;
+
+    private volatile boolean stopping;
+    private boolean connected;
+    private Message previousMessage;
+
+    public OpenThermGatewaySocketConnector(OpenThermGatewayCallback callback, String ipaddress, int port) {
+        this.callback = callback;
+        this.ipaddress = ipaddress;
+        this.port = port;
+    }
+
+    @Override
+    public synchronized void stop() {
+        callback.log(LogLevel.Debug, "Stopping OpenThermGatewaySocketConnector");
+        try {
+            socket.close();
+        } catch (IOException ignore) {
+        }
+        stopping = true;
+    }
+
+    @Override
+    public void run() {
+        stopping = false;
+        connected = false;
+
+        try {
+            callback.log(LogLevel.Info,
+                    String.format("Connecting OpenThermGatewaySocketConnector to %s:%s", this.ipaddress, this.port));
+
+            callback.connecting();
+
+            socket = new Socket();
+            socket.connect(new InetSocketAddress(this.ipaddress, this.port), COMMAND_TIMEOUT);
+            socket.setSoTimeout(COMMAND_TIMEOUT);
+            writer = new PrintWriter(socket.getOutputStream(), true);
+            reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
+
+            connected = true;
+
+            callback.connected();
+
+            callback.log(LogLevel.Debug, "OpenThermGatewaySocketConnector connected");
+
+            sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintReport, "A"));
+            // Set the OTGW to report every message it receives and transmits
+            sendCommand(GatewayCommand.parse(GatewayCommandCode.PrintSummary, "0"));
+
+            while (!stopping && !Thread.currentThread().isInterrupted()) {
+                String message = reader.readLine();
```

Comment:
Won't this wait forever, if no response from the gateway is received?

Reply:
That is possible, which is one of the reasons it runs on a separate thread.